### PR TITLE
EKS wait for management cluster and pinniped to be healthy and set kubeconfig when wait_for_kubeconfig is set to true

### PR DESCRIPTION
**Pull request by sreenivasmrpivot**
1. **What this PR does / why we need it**:
EKS wait for management cluster and pinniped to be healthy and set kubeconfig when wait_for_kubeconfig is set to true

2. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes # TMC-43744

3. **Additional information**

4. **Special notes for your reviewer**

### CODE REVIEW

File: internal/resources/ekscluster/data_source.go [hunk 1]
```diff
@@ -79,6 +82,36 @@ func dataSourceTMCEKSClusterRead(ctx context.Context, d *schema.ResourceData, m
 			return true, nil
 		}
 
+		if isWaitForKubeconfig(d) {
+			clusFullName := &clustermodel.VmwareTanzuManageV1alpha1ClusterFullName{Name: resp.EksCluster.Spec.AgentName, OrgID: clusterFn.OrgID, ManagementClusterName: "eks", ProvisionerName: "eks"}
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source.go:86]: nit: break this into multiple lines for readability.

File: internal/resources/ekscluster/data_source.go [hunk 2]
```diff
@@ -79,6 +82,36 @@ func dataSourceTMCEKSClusterRead(ctx context.Context, d *schema.ResourceData, m
...
+			if !isManagemetClusterHealthy(clusterResp) {
+				log.Printf("[DEBUG] waiting for cluster(%s) to be in Healthy status", clusterFn.Name)
+				return true, nil
+			}
+
+			fn := &configModels.VmwareTanzuManageV1alpha1ClusterFullName{
+				ManagementClusterName: "eks",
+				ProvisionerName:       "eks",
+				Name:                  resp.EksCluster.Spec.AgentName,
+				OrgID:                 "",
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source.go:103]: nit: remove empty assignments

File: internal/resources/ekscluster/data_source.go [hunk 3]
```diff
@@ -79,6 +82,36 @@ func dataSourceTMCEKSClusterRead(ctx context.Context, d *schema.ResourceData, m
 			return true, nil
 		}
 
+		if isWaitForKubeconfig(d) {
+			clusFullName := &clustermodel.VmwareTanzuManageV1alpha1ClusterFullName{Name: resp.EksCluster.Spec.AgentName, OrgID: clusterFn.OrgID, ManagementClusterName: "eks", ProvisionerName: "eks"}
+			clusterResp, err := config.TMCConnection.ClusterResourceService.ManageV1alpha1ClusterResourceServiceGet(clusFullName)
+			// nolint: wsl
+			if err != nil {
+				log.Printf("Unable to get Tanzu Mission Control cluster entry, name : %s", clusterFn.Name)
+				return true, nil
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source.go:91]: Return the error too. In case this fails with the timeout, we can see the error. Also, log the error.

File: internal/resources/ekscluster/data_source.go [hunk 4]
```diff
@@ -79,6 +82,36 @@ func dataSourceTMCEKSClusterRead(ctx context.Context, d *schema.ResourceData, m
...
+				ProvisionerName:       "eks",
+				Name:                  resp.EksCluster.Spec.AgentName,
+				OrgID:                 "",
+			}
+			resp, err := config.TMCConnection.KubeConfigResourceService.KubeconfigServiceGet(fn)
+
+			if kubeConfigReady(err, resp) {
+				if err = d.Set(kubeconfigKey, resp.Kubeconfig); err != nil {
+					log.Printf("Failed to set Kubeconfig")
+					return true, nil
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source.go:110]: If setting the kubeconfig fails, we don't need to retry, it's probably non-recoverable error.

File: internal/resources/ekscluster/data_source.go [hunk 5]
```diff
@@ -79,6 +82,36 @@ func dataSourceTMCEKSClusterRead(ctx context.Context, d *schema.ResourceData, m
...
+
+			fn := &configModels.VmwareTanzuManageV1alpha1ClusterFullName{
+				ManagementClusterName: "eks",
+				ProvisionerName:       "eks",
+				Name:                  resp.EksCluster.Spec.AgentName,
+				OrgID:                 "",
+			}
+			resp, err := config.TMCConnection.KubeConfigResourceService.KubeconfigServiceGet(fn)
+
+			if kubeConfigReady(err, resp) {
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source.go:107]: There is no `else` for this if. So, if the kubeconfig is not ready, it will reach the end of the function which marks this operation as successful.

File: internal/resources/ekscluster/data_source.go (hunk 4 above)
Comment (asutoshpalai) [reply internal/resources/ekscluster/data_source.go:110]: You also need to return the error so it's propagated up.

File: internal/resources/ekscluster/data_source.go [hunk 6]
```diff
@@ -79,6 +82,36 @@ func dataSourceTMCEKSClusterRead(ctx context.Context, d *schema.ResourceData, m
...
+				return true, nil
+			}
+
+			fn := &configModels.VmwareTanzuManageV1alpha1ClusterFullName{
+				ManagementClusterName: "eks",
+				ProvisionerName:       "eks",
+				Name:                  resp.EksCluster.Spec.AgentName,
+				OrgID:                 "",
+			}
+			resp, err := config.TMCConnection.KubeConfigResourceService.KubeconfigServiceGet(fn)
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source.go:105]: Check the error here.

File: internal/resources/ekscluster/data_source_test.go [hunk 7]
```diff
@@ -62,3 +60,45 @@ func TestNodepoolPosMap(t *testing.T) {
...
+				Cluster: &clustermodel.VmwareTanzuManageV1alpha1ClusterCluster{
+					Status: &clustermodel.VmwareTanzuManageV1alpha1ClusterStatus{
+						Health: clustermodel.NewVmwareTanzuManageV1alpha1CommonClusterHealth(clustermodel.VmwareTanzuManageV1alpha1CommonClusterHealthHEALTHY),
+					},
+				},
+			},
+			response: true,
+			err:      nil,
+		},
+	}
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source_test.go:95]: Add a test case for error?

**Review by asutoshpalai — CHANGES_REQUESTED** [review]
Can you also please add tests to `internal/resources/ekscluster/resource_ekscluster_test.go`?

File: internal/resources/ekscluster/data_source.go [hunk 8]
```diff
@@ -79,6 +82,52 @@ func dataSourceTMCEKSClusterRead(ctx context.Context, d *schema.ResourceData, m
...
+				ProvisionerName:       "eks",
+			}
+			clusterResp, err := config.TMCConnection.ClusterResourceService.ManageV1alpha1ClusterResourceServiceGet(clusFullName)
+			// nolint: wsl
+			if err != nil {
+				log.Printf("Unable to get Tanzu Mission Control cluster entry, name : %s, error :  %s", clusterFn.Name, err.Error())
+				return true, err
+			}
+
+			mgmtClusterHealthy, err := isManagemetClusterHealthy(clusterResp)
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source.go:99]: nit: `isManagemetClusterHealthy` is not a management cluster, it's just a (workload) cluster.

File: internal/resources/ekscluster/data_source.go [hunk 9]
```diff
@@ -79,6 +82,52 @@ func dataSourceTMCEKSClusterRead(ctx context.Context, d *schema.ResourceData, m
...
+			if err != nil {
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source.go:100]: nit: since this `if` and its `else if` have the same body, why not merge it?

Also, maybe returning the from that method error is not necessary, just bool should be fine?

File: internal/resources/ekscluster/data_source_test.go [hunk 10]
```diff
@@ -1,6 +1,3 @@
-//go:build ekscluster
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source_test.go:1]: Why this change? Also why was it here in the first place?

File: internal/resources/ekscluster/data_source.go [hunk 11]
```diff
@@ -124,6 +173,22 @@ func dataSourceTMCEKSClusterRead(ctx context.Context, d *schema.ResourceData, m
...
+	}
+
+	if *cluster.Cluster.Status.Health == clustermodel.VmwareTanzuManageV1alpha1CommonClusterHealthHEALTHY {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func kubeConfigReady(err error, resp *configModels.VmwareTanzuManageV1alpha1ClusterKubeconfigGetKubeconfigResponse) bool {
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/data_source.go:188]: Since the error is checked before calling this method, do we need to check it again?

**Review by asutoshpalai — DISMISSED** [review]
Minor code style comments.

File: internal/resources/ekscluster/data_source.go (hunk 1 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source.go]: @asutoshpalai  fixed it

File: internal/resources/ekscluster/data_source.go (hunk 2 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source.go]: @asutoshpalai  fixed it

File: internal/resources/ekscluster/data_source.go (hunk 3 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source.go]: @asutoshpalai  fixed it

File: internal/resources/ekscluster/data_source.go (hunk 4 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source.go]: @asutoshpalai  fixed it

File: internal/resources/ekscluster/data_source.go (hunk 5 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source.go:120]: @asutoshpalai  fixed it

File: internal/resources/ekscluster/data_source.go (hunk 6 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source.go:113]: @asutoshpalai  fixed it

File: internal/resources/ekscluster/data_source_test.go (hunk 7 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source_test.go:106]: @asutoshpalai  fixed it

File: internal/resources/ekscluster/data_source_test.go (hunk 10 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source_test.go:1]: @asutoshpalai If we have it, test cases in the file are skipped. 
I am not sure why it was added @ramya-bangera  might have some insight

File: internal/resources/ekscluster/data_source.go (hunk 8 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source.go]: @asutoshpalai fixed it

File: internal/resources/ekscluster/data_source.go (hunk 9 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source.go]: @asutoshpalai fixed it

File: internal/resources/ekscluster/data_source.go (hunk 11 above)
Comment (sreenivasmrpivot) [reply internal/resources/ekscluster/data_source.go]: @asutoshpalai fixed it

**Review by asutoshpalai — APPROVED** [review]

**Review by ramya-bangera — APPROVED** [review]
